src/cache.rs

```diff
@@ -0,0 +1,283 @@
+// Copyright 2016 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+use std::collections::BTreeMap;
+use std::cmp::Ord;
+use std::fmt::Debug;
+
+use rusqlite;
+
+use errors::*;
+use mentat_core::{
+    HasSchema,
+    KnownEntid,
+    NamespacedKeyword,
+    Schema,
+    TypedValue,
+};
+
+use query::{
+    q_once,
+    IntoResult,
+    QueryInputs,
+    Variable,
+};
+
+pub enum CacheAction {
+    Add,
+    Remove,
+}
+
+pub trait ValueProvider<K, V>: Sized {
+    fn values_for_attribute<'sqlite, 'schema>(&mut self, sqlite: &'sqlite rusqlite::Connection, schema: &'schema Schema, attribute: &K) -> Result<V>;
+}
+
+pub trait Cacheable {
+    type Key;
+    type Value;
+    type ValueProvider;
+
+    fn new(value_provider: Self::ValueProvider) -> Self;
+    fn contains_key(&self, key: &Self::Key) -> bool;
+    fn add<'sqlite, 'schema>(&mut self,
+                             sqlite: &'sqlite rusqlite::Connection,
+                             schema: &'schema Schema,
+                             key: Self::Key) -> Result<()>;
+    fn remove(&mut self, key: &Self::Key) -> Result<Self::Value>;
+    fn get(&mut self, key: &Self::Key) -> Result<&Self::Value>;
+}
+
+pub struct EagerCache<K, V, VP> where K: Ord, VP: ValueProvider<K, V> {
+    cache: BTreeMap<K, V>,
+    value_provider: VP,
+}
+
+impl<K, V, VP> Cacheable for EagerCache<K, V, VP> where K: Ord + Clone + Debug, V: Clone, VP: ValueProvider<K, V> {
+    type Key = K;
+    type Value = V;
+    type ValueProvider = VP;
+
+    fn new(value_provider: Self::ValueProvider) -> Self {
+        EagerCache {
+            cache: BTreeMap::new(),
+            value_provider,
+        }
+    }
+
+    fn contains_key(&self, key: &Self::Key) -> bool {
+        self.cache.contains_key(key)
+    }
+
+    fn add<'sqlite, 'schema>(&mut self,
+                                            sqlite: &'sqlite rusqlite::Connection,
+                                            schema: &'schema Schema,
+                                            key: Self::Key) -> Result<()> {
+        // fetch results and add to cache
+        let values = self.value_provider.values_for_attribute(sqlite, schema, &key)?;
+        self.cache.insert(key.clone(), values);
```

Comment:
This is not how I thought this would work.

If I'm reading this correctly, you have a map from attribute to T, where T is a map from entity to array of values?

I expect a handful of attributes to be cached. Why not simplify the cache to be the entity-value map, and then have one cache per attribute, and manage those elsewhere? That encapsulates the caching behavior and gets rid of the confusing meaning of 'key' as 'attribute'.